tools/codegen/gen_backend_stubs.py

```diff
@@ -171,6 +171,23 @@ def create_decl(f: NativeFunction) -> str:
 """
     assert missing_kernels_err_msg == "", missing_kernels_err_msg
 
+    # Additionally, provide a nice error if a user tries to register a composite op with no derivative formula to a backend key.
+    # Otherwise, autograd will silently stop working for their op.
+    for f in native_functions:
+        if backend_indices[backend_key].has_kernel(f) and f.has_composite_implicit_autograd_kernel:
+            # At this point, we know that the vendor provided a kernel for an op that also has a CompositeImplicitAutograd kernel.
+            # What we really want to error on, though, is if that op doesn't have a derivative formula registered in-tree.
+            # We can guess that based on whether or not the op has ANY in-tree non-composite kernels.
+            # Technically this has a small chance of false negatives, but it's easier than re-parsing derivatives.yaml here.
+            has_non_composite_in_tree_backend = any(idx.has_kernel(f) for idx in backend_indices.values() if not idx.external)
+            assert not has_non_composite_in_tree_backend, \
+                f"\nFound an entry for '{f.func.name}' in the yaml. " \
+                "This operator is composite, which means that by default it will decompose into base operators. " \
+                "If decomposing is fine, then you can simply remove the entry from the yaml file. " \
+                "If you wish to provide an explicit kernel directly for this composite op, you can do so by:\n " \
+                "(a) moving the operator name under the 'autograd' entry in the yaml file\n " \
+                "(b) writing a custom Autograd Function for the operator"
```

Comment:
I have one open:  is it possible to first get a list of register-able OPs by scanning the backend_indices with in tree backends except DispatchKey.CompositeImplicitAutograd and DispatchKey.CompositeExplicitAutograd, and then check if the OPs registered by external backend are in the list or not.

The reason is that I want to get the list: the register-able OPs that we have not registered in our backend.  This would be easy if we have the list of all register-able OPs.

Reply:
That depends on what you're looking for.

If you're interested in "what are all of the registerable ops", you can just look at `aten/src/ATen/native/native_functions.yaml`, which has the complete list.

The functions in the `DispatchKey.CompositeImplicitAutograd` and `DispatchKey.CompositeExplicitAutograd` probably aren't what you want, because those are ops that have composite implementations, a custom backend like XPU would get them for free (they would decompose into base ops, which you would then have to implement).

Reply:
yes, we can get it from native_functions.yaml, my thinking was to make this script provide a new feature (list un-registered OPs) by adding a little effort since we already has similar logic here.  It doesn't matter if we decide not to provide such OPs in this script, I'm open to write another one.

Reply:
There's actually a more programmatic way you can figure out which ops have registrations.
```
import torch  # this runs all registrations to the dispatcher in core
import torch_ipex  # assuming this runs dispatcher registrations for XPU

# This prints EVERY operator that's registered to the dispatcher
torch._C._dispatch_print_registrations_for_dispatch_key()

# This prints every op with a CompositeImplicitAutograd registration
torch._C._dispatch_print_registrations_for_dispatch_key('CompositeImplicitAutograd')

# This prints every op with an XPU registration
torch._C._dispatch_print_registrations_for_dispatch_key('XPU')
```

Reply:
hi @bdhirsh , sorry for the late response, I tried "torch._C._dispatch_print_registrations_for_dispatch_key()" and found that all the ops in native_functions.yaml are returned, including ops DEPRECATED, ops with manual_cpp_binding=True, and other ops not registered with CUDA/CPU in native_functions.yaml.

```
# DEPRECATED. DO NOT USE
- func: _cast_Byte(Tensor self, bool non_blocking=False) -> Tensor
  variants: function

- func: _backward(Tensor self, Tensor[] inputs, Tensor? gradient=None, bool? retain_graph=None, bool create_graph=False) -> ()
  manual_cpp_binding: True
  variants: method

- func: _fw_primal(Tensor(a) self, int level) -> Tensor(a)
  variants: method
  dispatch:
    CompositeExplicitAutograd: _fw_primal

- func: _make_dual(Tensor(a) primal, Tensor tangent, int level) -> Tensor(a)
  variants: function

- func: refine_names(Tensor(a) self, Dimname[] names) -> Tensor(a)
  variants: method

- func: _sobol_engine_initialize_state_(Tensor(a!) self, int dimension) -> Tensor(a!)

- func: silu_backward(Tensor grad_output, Tensor self) -> Tensor
  structured_delegate: silu_backward.grad_input
  python_module: nn
  dispatch:
    CompositeImplicitAutograd: math_silu_backward

```

From an external backend's viewpoint, we want to get all the OPs we do need to register, any guidance for us to quickly filter the ops, thanks.